### PR TITLE
buildRustCrateTests: Fix link order test on darwin

### PR DESCRIPTION
**Pull request by andir**
###### Motivation for this change

Fixes the tests on Darwin that I broke by merging #83379 
cc @kolloch 

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS (does that have a sandbox?)

### CODE REVIEW

File: pkgs/build-support/rust/build-rust-crate/test/default.nix [hunk 1]
```diff
@@ -284,12 +293,16 @@ let
...
+          compile = name: text: let
+            src = writeTextFile {
               name = "${name}-src.c";
               inherit text;
-            }}
+            };
+          in runCommandCC name {} ''
+            mkdir -p $out/lib
+            cc -shared \
+              ${lib.optionalString stdenv.isDarwin "-undefined dynamic_lookup"} \
```
Comment (flokli) [inline pkgs/build-support/rust/build-rust-crate/test/default.nix]: I was part of that debugging session - surprisingly, it worked with a linux clang (which is why this probes for `stdenv.isDarwin` and not `stdenv.cc.isClang`.
Maybe we want to drop a comment about that in this file, for posterity.

**Review by flokli — APPROVED** [review]

**Review by LnL7 — APPROVED** [review]
Looks reasonable, assuming there's a reason to not properly link the example `libb` against `liba`.

**Comment by andir** [comment]
On 15:47 28.03.20, Daiderd Jordan wrote:
> @LnL7 approved this pull request.
> 
> Looks reasonable, assuming there's a reason to not properly link the example `libb` against `liba`.

Yeah, that was intentional since we are trying to simulate a behaviour
of cargo + rustc.

I am a bit baffled that such measures are required on Darwin. On Linux
it just works out of the box. Shouldn't this maybe be handled at a more
generic level within nixpkgs?

**Comment by LnL7** [comment]
Looks like there are a couple of cases out there, but I've not encountered that before. Presumably things that need this already handle it in their build system making it something we don't really have to worry about.

**Pull request merged**